File: src/skills/custom/logging-to-php-api/scripts/LettaLogger.ts
/**
 * LettaLogger — sends structured log entries to the americansjewelry.com PHP API.
 *
 * Each class gets one envelope row in monitored_objects keyed by objectViewId
 * (convention: ClassName_Year, e.g. "LettaClient_2026").
 *
 * objectData is a JSON-encoded array of log entries that is fully replaced on
 * every flush.  Logging never throws — failures are silently swallowed so they
 * cannot crash the host object.
 */

const BASE_URL =
  "https://americansjewelry.com/libraries/local-php-api/index.php/object";
const TIMEOUT_MS = 5000;

export type LogStatus = "green" | "yellow" | "red";

interface LogEntry {
  timestamp: string;
  method: string;
  event: string;
  status: LogStatus;
  data?: unknown;
}

export class LettaLogger {
  private readonly objectViewId: string;
  private entries: LogEntry[] = [];

  constructor(objectViewId: string) {
    this.objectViewId = objectViewId;
  }

  // --------------------------------------------------------------------
  // Public API
  // --------------------------------------------------------------------

  log(
    method: string,
    event: string,
    data?: unknown,
    status: LogStatus = "green",
  ): void {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      method,
      event,
      status,
      ...(data !== undefined ? { data } : {}),
    };
    this.entries.push(entry);
    this.flush().catch(() => {}); // fire-and-forget, never throws
  }

  // --------------------------------------------------------------------
  // Internal
  // --------------------------------------------------------------------

  private async flush(): Promise<void> {
    const payload = JSON.stringify({
      object_view_id: this.objectViewId,
      object_data: JSON.stringify(this.entries),
    });

    const inserted = await this.post("/insert", payload);
    if (!inserted) {
      await this.post("/update", payload);
    }
  }

  /** Returns true if insert succeeded (response contains "insert success"). */
  private async post(path: string, payload: string): Promise<boolean> {
    try {
      const res = await fetch(BASE_URL + path, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: payload,
        signal: AbortSignal.timeout(TIMEOUT_MS),
      });
      const text = await res.text();
      return text.includes("insert success");
    } catch {
      return false;
    }
  }
}

File: src/skills/custom/logging-to-php-api/scripts/letta_logger.py
"""
LettaLogger — sends structured log entries to the americansjewelry.com PHP API.

Each class gets one envelope row in monitored_objects keyed by object_view_id
(convention: ClassName_Year, e.g. "ProviderManager_2026").

object_data is a JSON-encoded list of log entries that is fully replaced on
every flush.  Logging never raises — failures are silently swallowed so they
cannot crash the host object.
"""

import json
import urllib.request
import urllib.error
from datetime import datetime
from typing import Any, Literal

_BASE_URL = "https://americansjewelry.com/libraries/local-php-api/index.php/object"
_TIMEOUT = 5  # seconds

Status = Literal["green", "yellow", "red"]


class LettaLogger:
    """Append-and-flush logger that upserts to the PHP monitored_objects API."""

    def __init__(self, object_view_id: str) -> None:
        self.object_view_id = object_view_id
        self._entries: list[dict] = []

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def log(
        self,
        method: str,
        event: str,
        data: Any = None,
        status: Status = "green",
    ) -> None:
        """Append a log entry and flush to the API (fire-and-forget)."""
        entry: dict = {
            "timestamp": datetime.now().isoformat(),
            "method": method,
            "event": event,
            "status": status,
        }
        if data is not None:
            entry["data"] = data
        self._entries.append(entry)
        self._flush()

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _flush(self) -> None:
        """Upsert current entries to the PHP API (insert-first strategy)."""
        payload = json.dumps({
            "object_view_id": self.object_view_id,
            "object_data": json.dumps(self._entries),
        }).encode("utf-8")

        if self._post("/insert", payload):
            return  # new row created
        self._post("/update", payload)  # row already existed

    def _post(self, path: str, payload: bytes) -> bool:
        """POST payload to BASE_URL+path.  Returns True on HTTP 2xx response."""
        try:
            req = urllib.request.Request(
                _BASE_URL + path,
                data=payload,
                headers={
                    "Content-Type": "application/json",
                    "Accept": "application/json, text/plain, */*",
                    "User-Agent": "Mozilla/5.0 (compatible; LettaLogger/1.0)",
                },
                method="POST",
            )
            with urllib.request.urlopen(req, timeout=_TIMEOUT) as resp:
                resp.read()  # drain response body
                return True  # urlopen raises on HTTP errors; reaching here means 2xx
        except Exception:
            return False
